### ISSUE TITLE
setup3.py unusable / missing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6228]
|<img alt="Arfrever's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Arfrever reported|
|-|-|
|Trac ID|trac#6228|
|Type|defect|
|Created|2012-12-26 22:05:27Z|
|Branch|https://github.com/twisted/twisted/tree/standalone-setup3-6228-2|

Twisted-12.3.0.tar.bz2 contains unusable setup3.py script:
```
$ python3.3 setup3.py
Traceback (most recent call last):
  File "setup3.py", line 24, in <module>
    exec(open("admin/_twistedpython3.py").read(), ported)
FileNotFoundError: [Errno 2] No such file or directory: 'admin/_twistedpython3.py'
```

`TwistedCore-12.3.0.tar.bz2`, `TwistedNames-12.3.0.tar.bz2` and `TwistedWeb-12.3.0.tar.bz2` do not contain setup3.py at all.

[admin/_twistedpython3.py](https://twistedmatrix.com/trac/browser/tags/releases/twisted-12.3.0/admin/_twistedpython3.py) implies that other tarballs do not need setup3.py.

<details><summary>Searchable metadata</summary>

```
trac-id__6228 6228
type__defect defect
reporter__Arfrever Arfrever
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_standalone_setup3_6228_2 branches/standalone-setup3-6228-2
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1356559527000000 1356559527000000
changetime__1370105690000000 1370105690000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I suspect we'll never get to providing `setup3.py` for each/any of the subproject tarballs.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @therve|
|-|-|

Release automation bug?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

It seems so. admin/ is not included in the tarballs for good reasons. Maybe we should generate a new setup3.py when creating them. Or probably just when creating the main tarball.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [36857]) Branching to 'generate-setup3-6228'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

The attached branch now generates a setup3.py file in the main tarball.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

Wouldn't it be simpler to move `_twistedpython3.py`, perhaps with a new name, to somewhere other than `admin/`, so it can be included? Or even have the list of modules in `setup3.py` itself, and have the Python 3 test runner script import `setup3`?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

It may be. Is it you review? :)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set owner to @therve|
|-|-|

Sure; but feel free to explain why current solution is better, if you think it is.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

If we are going to be discussing this for a while, I propose that we just remove `setup3.py` from the release tarballs entirely for the time being, since that will fix the issue of shipping a tool that just doesn't work.

I agree with itamar that we ought to just move the `_twistedpython3` module so it's importable.  However, I'd like to expand on that a bit, since this seems to be a possible slide back into the bad old days of the 'submodule split' where we put too much pressure on our release tooling infrastructure and introduced lots of pointless variances between twisted-as-developed and twisted-as-packaged.

It seems to me that there is no clear reason to keep `_twistedpython3.py`, or for that matter `run-python3-tests`, out of the tarball.  These are both temporary tools which should go away eventually, but if someone is going to be packaging our supported subset of Python 3, I want it to be as easy as possible to convince them to package it in the way that we suggest.

As I understand it, the purpose of the `admin/` directory is to contain tools that neither users nor packagers need to run themselves, to avoid cluttering the tarball with them.  Over the last few years we've nearly emptied that directory entirely, which is generally a good thing, since "internal" procedures don't make a whole lot of sense on a public open source project, especially when it is published on multiple platforms which each have their own build and packaging systems.  Eventually, I'd like to eliminate the directory entirely.  Arguably the reason that `update-documentation.py` is in there is because these packagers ought to be downloading the documentation that we have generated ourselves with up-to-the-minute and never-quite-documented versions of our documentation generation tools, and are therefore discouraged from doing the generation themselves.  That seems reasonable, because there's a clear alternative, but it would be better if we just _did_ have a clear specification of what documentation tools it depended upon and a reliable explanation of how to run them.  `twisted.spec` is in there because it's platform-specific and there's not really any other clear place to put it.  Maybe it ought to be just removed; it doesn't see a lot of maintenance and we don't use it as part of our official release process.

I don't want anyone to try to force all of Twisted to work on python 3; I want the level of support to be made clear by the fact that only a few modules show up in a `python3 setup3.py sdist` or `python3 setup3.py install`.  Already some packagers have been in touch after some confusion as to how they're supposed to get it to work, attempting to run 2to3 over our codebase.  If such versions are broadly distributed, we will potentially start getting lots of incredibly confusing bug reports from people who tried to use their packaged Twisted on python 3.

We're still fighting the battle with packagers that we want them to run our unit tests on their platforms as part of their QA process.  The fact that 'trial' isn't packaged yet, and therefore they need to run a different tool, is a technical necessity, but by obscuring `run-python3-tests`, we reinforce the already-common misapprehension that our testing tools are internal and should not be run by packagers as part of an integration test.

The solution implemented in the branch, using a template file, introduces a bunch of additional complexity which only makes sense if there's a sound technical reason to keep `_twistedpython3.py` hidden from people who download the tarball.  However, if someone wants to document the reason for that, I don't have any other opposition to the idea that it ought to be done that way.

There are still some practical problems with the current implementation though, so here's an _actual_ code review in case therve comes back to this or someone else wants to continue with the current branch:

  1. While `test_release` checks to see that some values get inserted into the template, nothing actually checks the existing template, even to make sure it's syntactically valid Python.  This seems like pretty light test coverage.
  2. There's no affordance in the release process to do this type of testing, because [the release process documentation](https://twistedmatrix.com/trac/wiki/ReleaseProcess) doesn't even mention running `setup.py` on python 2, not to mention testing the newly-generated `setup3.py` for python 3.
  3. There's no affordance in the development process to even accidentally do this type of testing, because developers will be running the bundled-in-version-control `setup3.py`, which bears no resemblance to the one that is generated by the automation tools - which is arguably the problem that this bug is about in the first place.

Finally, I would like to note that I don't think putting the list of modules into `setup3.py` itself, as itamar suggested, would be a good idea.  `import setup3` is fraught with peril, as is `import setup`, because someone doing python3 development might easily have a dozen `setup3.py`'s on their `sys.path`, and diagnosing the error condition when they pick the wrong one - especially if they implicitly pick the wrong one in the bowels of some thing that uses setuptools - can be very tricky.  It's been a long time since I've had to diagnose an error like that and I'd like it to be even longer :).  Similarly, loading the `setup3.py` module with `exec()` or similar could be problematic because then things would need to compute the correct path to that module and then the code to do that path computation would need its own tests, and I don't see much of an advantage to either of those approaches over just packaging a regular module and letting the overly-complex-but-at-least-mostly-understood logic of `import` be the source of any deployment headaches.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @glyph|
|-|-|

OK, I'm not going to read that giant blob of text again, sorry. I saw some complaints though. My suggestions to move forward:
 * Put the list of supported stuff in setup3.py
 * Use execfile in admin/run-python3-tests to get the list of test modules
 * Remove admin/_twistedpython3.py

Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_13):
> OK, I'm not going to read that giant blob of text again, sorry.

No, I'm sorry, I really should have edited that down or at least broken it up.

> I saw some complaints though. My suggestions to move forward:
>  * Put the list of supported stuff in setup3.py
>  * Use execfile in admin/run-python3-tests to get the list of test modules
>  * Remove admin/_twistedpython3.py
> 
> Thoughts?

Instead of a big blob of text, how about this:

"Yes".

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

(In [37794]) Branching to 'standalone-setup3-6228'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve removed owner|
|-|-|

Alright, I've done just that. I also fixed the list of modules to be able to test the tarball (there were some missing dependencies). Please review!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince commented|
|-|-|

There should buildbot to test that installing works. I guess something like the easy_install builders?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_20"></a>|@wallrj set owner to @therve<br>@wallrj set status to `new`|
|-|-|

## Code Review
Hi therve,

Thanks for the branch.

Here's how I tested:
 1. Installed virtualenvs for trunk and branch installs (eg
    mkvirtualenv --python=python3.3 --no-site-packages
    SetupTwisted3.trunk)
 2. I attempted a Python3 install from a clean svn export of trunk
    (after renaming the admin dir) and confirmed the original bug.
 3. svn export standalone-setup-6228 branch, renamed admin and ran
    python3.3 setup3.py install.
 4. Installation completed without errors.
 5. Ran a twisted names lookup from the python3.3 command line - which
    worked.
    ```
    (SetupTwisted3.standalone-setup3-6228)[richard@zorin ~]$ python
    Python 3.3.0 (default, Sep 29 2012, 22:07:38)
    [GCC 4.7.2 20120921 (Red Hat 4.7.2-2)] on linux
    Type "help", "copyright", "credits" or "license" for more information.
    >>> from twisted.internet import reactor
    >>> from twisted.names import client
    >>> import sys
    >>> d = client.lookupAddress('www.example.com').addCallback(lambda res: sys.stdout.write(repr(res)+'\n')).addBoth(lambda res: reactor.stop())
    >>> reactor.run()
    ([<RR name=www.example.com type=A class=IN ttl=86160s auth=False>], [], [])
    ```


Please answer or address the following:
 1. Missing news file.
 2. [branches/standalone-setup3-6228/setup3.py](branches/standalone-setup3-6228/setup3.py)
    1. unlike setup.py - this doesn't contain a requirements section -
       why not?
       1. So when I tried to renaming setup3.py to setup.py and running
          "pip install ." the install didn't automatically download and
          install zope.interface.
       2. Should the requirements be added to
          source:branches/standalone-setup3-6228/twisted/python/dist.py ?
 3. [branches/standalone-setup3-6228/admin/run-python3-tests](branches/standalone-setup3-6228/admin/run-python3-tests)
    1. I read that it's better to pass a code object to exec, so that
       it can print meaningful filenames in stacktraces.
       1. Should we do that here?
          http://stackoverflow.com/a/437857/919436
       2. See twisted.python.compat.execfile (I know it can't be used
          here, but it also has a compile step.)
 4. Raise a twisted-buildbot ticket so we don't forget Tom Prince's
    suggestion (above) about a python3 install builder.
 5. Glyph said (above) "...Similarly, loading the setup3.py module with exec()
    or similar could be problematic because then things would need to
    compute the correct path to that module and then the code to do
    that path computation would need its own tests"

    What is your response to that?


Build results look fine:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/standalone-setup3-6228

If you address or answer the points above I'd be happy for you to
merge this branch without further review.

But perhaps it would be worth asking someone on #twisted-dev with more
Python3 experience to review the branch too.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve commented|
|-|-|

(In [38603]) Branching to 'standalone-setup3-6228-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

Thanks for the review. Some comments:

Replying to [rwall](#note_20):

>  1. Missing news file.

Fixed.

>  2. [branches/standalone-setup3-6228/setup3.py](branches/standalone-setup3-6228/setup3.py)
>     1. unlike setup.py - this doesn't contain a requirements section -
>        why not?
>        1. So when I tried to renaming setup3.py to setup.py and running
>           "pip install ." the install didn't automatically download and
>           install zope.interface.
>        2. Should the requirements be added to
>           [branches/standalone-setup3-6228/twisted/python/dist.py](branches/standalone-setup3-6228/twisted/python/dist.py) ?

Postponed.

>  3. [branches/standalone-setup3-6228/admin/run-python3-tests](branches/standalone-setup3-6228/admin/run-python3-tests)
>     1. I read that it's better to pass a code object to exec, so that
>        it can print meaningful filenames in stacktraces.
>        1. Should we do that here?
>           http://stackoverflow.com/a/437857/919436
>        2. See twisted.python.compat.execfile (I know it can't be used
>           here, but it also has a compile step.)

Fixed.

>  4. Raise a twisted-buildbot ticket so we don't forget Tom Prince's
>     suggestion (above) about a python3 install builder.

I'll do that.

>  5. Glyph said (above) "...Similarly, loading the setup3.py module with exec()
>     or similar could be problematic because then things would need to
>     compute the correct path to that module and then the code to do
>     that path computation would need its own tests"
> 
>     What is your response to that?

It's not as bad as he thought.

Merging.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve set status to `closed`|
|-|-|

(In [38607]) Merge standalone-setup3-6228-2

Author: therve
Reviewer: rwall
Fixes: [#6228](https://github.com/twisted/twisted/issues/6228)

Remove admin/_twistedpython3.py and put the list of ported modules in setup3.py
so that a source tarball can be installed with it.
